tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <ztest.h>
+#include <sys/rb.h>
+
+/* rbnode structure is embeddedable in user structure */
+struct container_node {
+	struct rbnode node;
+	int value;
+};
+
+/* Our "lessthan" is just the location of the struct */
+bool node_lessthan(struct rbnode *a, struct rbnode *b)
+{
+	return a < b;
+}
+
+/**
+ * @brief Test whether rbtree node struct is embedded
+ * in any user struct
+ *
+ * @details
+ * Initialize a user defined struct contains rbtree node
+ * Add into a rbtree
+ * Enumerate the rbtree node
+ *
+ * @ingroup lib_rbtree_tests
+ *
+ * @see RB_FOR_EACH(),RB_FOR_EACH_CONTAINER()
+ */
+void test_rbtree_container(void)
+{
+	int count = 0;
+	static struct rbtree test_tree_l;
+	struct container_node tree_node[10];
+	struct container_node *c_foreach_node;
+	struct rbnode *foreach_node;
+
+	memset(&test_tree_l, 0, sizeof(test_tree_l));
+	test_tree_l.lessthan_fn = node_lessthan;
+	memset(tree_node, 0, sizeof(tree_node));
+	for (unsigned int i = 0; i < 10; i++) {
+		tree_node[i].value = i;
+		rb_insert(&test_tree_l, &tree_node[i].node);
+	}
+
+	/*"for each" style iteration to verify node*/
+	RB_FOR_EACH(&test_tree_l, foreach_node) {
+		zassert_true(((struct container_node *)foreach_node)->value
+				== count, "RB_FOR_EACH failed");
+		count++;
+	}
+
+	count = 0;
+	/*"for each" style iteration to verify node*/
+	RB_FOR_EACH_CONTAINER(&test_tree_l, c_foreach_node, node) {
+		zassert_true(c_foreach_node->value == count,
+				"RB_FOR_EACH_CONTAINER failed");
+		count++;
+	}
+}
+
+static unsigned int next_rand_mod(unsigned int mod)
+{
+	static unsigned long long state = 123456789; /* seed */
+
+	state = state * 2862933555777941757ul + 3037000493ul;
+
+	return ((unsigned int)(state >> 32)) % mod;
+}
+
+static struct rbnode nodes[512];
+static int random_num[512] = {0};
+void random_generate(int size)
```

Comment:
I'm completely confused about what's happening here.  Are you trying to shuffle the integers between 0 and 511 into the random_num array such that they appear in a random order, but only once each?  If so, I don't think this code works correctly.  It's also needlessly O(N^2) in the size of the array.  Maybe try this instead (untested, but the basic idea)?:

```c


void shuffle_random(void)
{
    int n = ARRAY_SIZE(random_num);

    /* Populate values */
    for (int i = 0; i < n; i++) {
        random_num[i] = i;
    }

    /* Randomize each by swapping with one of the remaining ones */
    for (int i = 0; i < n; i++) {
        int j = next_rand_mod(n - i);
        int tmp = random_num[i];

        random_num[i] = random_num[j];
        random_num[j] = tmp;
    }
}
```

Reply:
The code is very good, but there are a few same numbers,as you know, it can't insert a tree using two same data.  If I need to use disorderly ordinal data, I think I can use this code instead and fix it.